### PR TITLE
sympify(set) gives FiniteSet

### LINKED ISSUE #10654: sympify(set) should give FiniteSet
```
In [14]: sympify({n})
Out[14]: set([n])
```

It should give FiniteSet(n), so that people can use that convenient syntax in set functions (like Union and so on). 

This is easy to fix. Just add it to the converter dictionary in sympify.py

### PR DESCRIPTION
**Pull request by AnishShah**
Fixes #10654

### CODE REVIEW

File: sympy/sets/sets.py [hunk 1]
```diff
@@ -1968,6 +1968,14 @@ def __lt__(self, other):
             raise TypeError("Invalid comparison of set with %s" % func_name(other))
         return self.is_proper_subset(other)
 
+    def __eq__(self, other):
```
Comment (AnishShah) [inline sympy/sets/sets.py]: I have added this because the old test was failing.
`FiniteSet(Rational(3, 10), Rational(1, 5)) == FiniteSet(0.3, 0.2)` was giving `False` before adding `__eq__`.
Let me know if this is fine?

**Comment by AnishShah** [comment]
What should be the output for:
`sympify(set())` - EmptySet()
`sympify({})`
Though in second case, it's a dict and not a set.

File: sympy/sets/sets.py [hunk 2]
```diff
@@ -1968,6 +1968,14 @@ def __lt__(self, other):
...
+        if not isinstance(other, Set):
+            raise TypeError("Invalid comparison of set with %s" % func_name(other))
```
Comment (asmeurer) [inline sympy/sets/sets.py]: This should give False, not TypeError.

File: sympy/sets/sets.py [hunk 3]
```diff
@@ -1968,6 +1968,14 @@ def __lt__(self, other):
...
+        return self.args == other.args
```
Comment (asmeurer) [inline sympy/sets/sets.py]: This is only valid if other is also a FiniteSet.

File: sympy/sets/sets.py (hunk 1 above)
Comment (asmeurer) [inline sympy/sets/sets.py]: I don't understand why your change would affect this.

File: sympy/sets/sets.py [hunk 4]
```diff
@@ -1968,6 +1968,14 @@ def __lt__(self, other):
...
+
+
+converter[set] = lambda x: FiniteSet(*x)
```
Comment (asmeurer) [inline sympy/sets/sets.py]: You can also add converter[frozenset]

File: sympy/core/tests/test_sympify.py [hunk 5]
```diff
@@ -502,3 +503,7 @@ def test_Range():
 
     assert sympify(builtin_range(10)) == Range(10)
     assert _sympify(builtin_range(10)) == Range(10)
+
+def test_sympify_set():
+    n = Symbol('n')
+    assert sympify({n}) == FiniteSet(n)
```
Comment (asmeurer) [inline sympy/core/tests/test_sympify.py]: If you use the set literal syntax here we won't be able to merge this until we remove Python 2.6 support in master (after the release).

File: sympy/sets/sets.py (hunk 1 above)
Comment (asmeurer) [inline sympy/sets/sets.py]: It seems the `__eq__` you are trying to write here is the same as `Basic.__eq__`, so I don't think it should be needed. I would try to figure out why this stopped working.

Comment (AnishShah) [inline sympy/sets/sets.py]: I just checked on live.sympy.org and master branch, `FiniteSet(Rational(3, 10), Rational(1, 5)) == FiniteSet(0.3, 0.2)`  does not work on master either.
And I think, it is because:

In [`Basic.__eq__`, it checks for `_hashable_content()`](https://github.com/sympy/sympy/blob/master/sympy/core/basic.py#L328) of both the operands.

And [`_hashable_content()` returns a `frozenset`](https://github.com/sympy/sympy/blob/master/sympy/sets/sets.py#L1941)

So basically, we are equating `frozenset([Rational(3, 10), Rational(1, 5)]) == frozenset([0.2, 0.3])` which is false

Comment (AnishShah) [inline sympy/sets/sets.py]: For `set/frozenset`, it does not call `__eq__` of its elements. Like `tuple` does.

```
In [2]: tuple([Rational(3, 10), Rational(1, 5)]) == tuple([0.3, 0.2])
Out[2]: True

In [3]: frozenset([Rational(3, 10), Rational(1, 5)]) == frozenset([0.3, 0.2])
Out[3]: False
```

Any idea, what would be a good way to solve this?

Comment (asmeurer) [inline sympy/sets/sets.py]: This looks like it's part of a bigger problem relating to equality of rationals and floats. I would just XFAIL the test and open a new issue for it (and remove this `__eq__`). The problem is that `Rational(3, 10) == S(0.3)` but `hash(3, 10) != hash(S(0.3))`. You can see my blog post about this to see why this is bad: https://asmeurer.github.io/blog/posts/what-happens-when-you-mess-with-hashing-in-python/.

**Comment by asmeurer** [comment]
Yeah, it's unfortunate, but `{}` is not the empty set. So SymPy should never printer the empty set as `{}`, to avoid the pitfall of using that in Python. `sympify({})` should probably create `Dict()` (that's another issue, though). I guess the type errors from combining dicts with sets should alert users to the mistake.

**Comment by AnishShah** [comment]
I have updated the PR with XFAIL and will open a new issue for Rational/Float hash  equality

**Comment by AnishShah** [comment]
@asmeurer Some tests related to stats module are failing. Any idea?

**Comment by asmeurer** [comment]
I don't know. You'll have to investigate.

File: sympy/stats/frv.py [hunk 6]
```diff
@@ -215,7 +215,7 @@ def domain(self):
 
     @property
     def density(self):
-        return self.args[0]
+        return self.args[1]
```
Comment (AnishShah) [inline sympy/stats/frv.py]: This should be `self.args[1]`, right? Since, `self.args[0]`  is for domain

**Comment by AnishShah** [comment]
I have updated the PR.

File: sympy/stats/frv.py [hunk 7]
```diff
@@ -319,7 +320,7 @@ def domain(self):
     @property
     @cacheit
     def _density(self):
-        return dict((frozenset(((self.symbol, val),)), prob)
+        return dict((sympify(frozenset(((self.symbol, val),))), prob)
```
Comment (jksuom) [inline sympy/stats/frv.py]: Perhaps it would be possible to use simply ``FiniteSet(...)`` instead of the combination ``sympify(frozenset(...))``.

File: sympy/stats/frv.py (hunk 6 above)
Comment (jksuom) [inline sympy/stats/frv.py]: I think you are right. Perhaps you could add a test to confirm this.

**Comment by jksuom** [comment]
I wish you would consider the suggestions above. Otherwise I think this looks good.

**Comment by AnishShah** [comment]
@jksuom Thanks! I have updated the PR

**Comment by jksuom** [comment]
Thanks! I think this is ready.
